crates/sui-core/src/authority/authority_per_epoch_store.rs

```diff
@@ -79,6 +84,15 @@ pub struct ExecutionIndicesWithHash {
     pub hash: u64,
 }
 
+// Data related to VM and Move execution and type layout
+pub struct ExecutionComponents {
+    /// Move native functions that are available to invoke
+    pub(crate) native_functions: NativeFunctionTable,
+    pub(crate) move_vm: Arc<MoveVM>,
+    // TODO: use strategies (e.g. LRU?) to constraint memory usage
+    pub module_cache: Arc<SyncModuleCache<ResolverWrapper<AuthorityStore>>>,
```

Comment:
the fact that the vm and cache are in `Arc` makes it very easy for code to hold on to them longer than they should.

I do not want to send you on a wild goose chase, but if it is is at all feasible to store them in directly, and return references to them instead, this would make a wide class of bugs impossible.

If the compiler makes that too difficult, one suggestion is we return a `Weak` and have the callers upgrade it each time they need to use it.

Reply:
not sure about this. In a sense this may get even "heavier" and become an `ArcSwap` to restart a VM within an epoch.
I am looking at that work soon and we'll see according to that